design/httpproxy-prefix-rewrite.md

```diff
@@ -0,0 +1,262 @@
+Distinguishing between these replacements requires specifying the routing prefix.
+
+* Prefix: This is the optional path prefix to match on.
+  If this is not specified, the replacement is applied for every routing match prefix.
+
+* Replacement: This is the string that replaces the path prefix. If a Prefix field is also specified, the replacement is only applied to the exact route prefix match.
+
+Note that the replacement prefix must be an exact match to exactly one routing prefix.
+
+### Future Extensions
+
+* Regex path rewrites can be added as a new field in the `PathRewritePolicy` struct.
+* Host header rewrites can be added as a new policy field in the `Route` struct, following similar semantics to this proposal.
+* Nothing in this proposal prevents adding prefix replacement specifications to non-leaf HTTPProxies.
+
+### Implementation Notes
+
+* ***Match type interaction.***
+  Envoy prefix rewriting can only be used with a prefix match route,
+  not with a regex match route. This means that Contour should
+  validate that prefix replacement may not be used in conjunction
+  with wildcard path matching.
+
+* ***Prefix fragility.***
+  Since a replacement prefix must be an exact match to a single
+  routing prefix, a rewrite configuration can be broken by changes
+  in parent HTTPProxies or by deployment ordering. We can mitigate
+  this by not making the lack of a Prefix match a hard validation
+  error. In this case, we would want richer status information to
+  indicate that some path rewrites are not being used.
+
+* ***Path component '/' ambiguity.***
+  This proposal does nothing to help users with the problem of
+  specifying trailing '/' characters correctly. If the prefix doesn't
+  end in a '/', then using a replacement of '/' can result in a
+  final path with a leading '//'.
+
+  *TODO(jpeach): What does Envoy do with an empty replacement?*
+
+* ***Multiple replacements.***
+  Rather than allowing a `ReplacementPrefix` slice, we could
+  obtain similar semantics by requiring the user to always
+  specify separate leaf proxies for each rendered prefix. This
+  locally simplifies the API, but arguably forces more complex
+  deployment topologies. If we can expose richer status
+  information, then I think the additional API complexity is
+  preferable.
+
+## Examples
+
+Rewrite a single prefix:
+
+```
+---
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: httpbin
+spec:
+  virtualhost:
+    fqdn: httpbin.jpeach.org
+  routes:
+  - conditions:
+    - prefix: /
+    services:
+    - name: httpbin
+      port: 80
+    pathRewrite:
+      replacePrefix:
+      - replacement: /my/great/replacement
+```
+
+Rewrite multiple prefixes to a single canonical prefix:
+
+```
+---
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: httpbin
+spec:
+  virtualhost:
+    fqdn: httpbin.jpeach.org
+  routes:
+  - conditions:
+    - prefix: /v1/
+    - prefix: /v2/
```

Comment:
This is not permitted at the moment, at most one prefix condition can be supplied per route.

Reply:
Fixed.